### PR TITLE
ISPN-13170 hibernate and spring guides

### PR DESCRIPTION
**Pull request by oraNod**
https://issues.redhat.com/browse/ISPN-13170 

Needs backport to 12.1.x for community site (required for stable doc links)

### CODE REVIEW

**Review by karesti — APPROVED** [review]

**Comment by karesti** [comment]
@oraNod LGTM don

File: documentation/src/main/asciidoc/stories/assembly_hibernate_cache.adoc [hunk 1]
```diff
@@ -0,0 +1,16 @@
+[id='hibernate-cache']
+:context: hibernate-cache
+= Using {brandname} as a second-level cache
+
+{brandname} implements the SPI that Hibernate exposes to allow integration with the second-level cache (2LC).
+This means you can use {brandname} to store data from `Session` or `EntityManager` (JPA) operations.
+
+link:https://docs.jboss.org/hibernate/orm/5.2/userguide/html_single/Hibernate_User_Guide.html#caching[Hibernate documentation]
```
Comment (danberindei) [inline documentation/src/main/asciidoc/stories/assembly_hibernate_cache.adoc:8]: Maybe we should link to 5.3, or even 6.0.

File: documentation/src/main/asciidoc/titles/hibernate/hibernate.asciidoc [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+:icons: font
+:toclevels: 3
+:numbered:
+:sectanchors:
+:stem: asciimath
+
+//Title
+= Hibernate Second-Level Caching
+
+{brandname} provides an interface so you can use it as a second-level cache (2LC) with Hibernate and store all sessions from applications using the same session factory.
```
Comment (danberindei) [inline documentation/src/main/asciidoc/titles/hibernate/hibernate.asciidoc:20]: Not sure about "provides an interface", I'd rather go with just "You can use {brandname} as ..."

File: documentation/src/main/asciidoc/topics/con_spring_cache_annotations.adoc [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+include::code_examples/SpringCachableExample.java[]
+----
+
+Any `Book` instances returned from `findBook(Integer bookId)` are stored in a cache named `books`, using `bookId` as the key.
+
+Note that "#bookId" is an expression in the link:http://static.springsource.org/spring/docs/current/spring-framework-reference/html/expressions.html[Spring Expression Language] that evaluates the `bookId` argument.
+
+[IMPORTANT]
+====
+If your application needs to reference entries in the cache directly, you should include the `key` attribute. Without this attribute, Spring generates a hash from the supplied method arguments to use as the cache key.
```
Comment (danberindei) [inline documentation/src/main/asciidoc/topics/con_spring_cache_annotations.adoc:24]: The default key generator does not use a hash of the arguments, it aggregates the arguments themselves in a SimpleKey:
https://docs.spring.io/spring-framework/docs/current/reference/html/integration.html#cache-annotations-cacheable-default-key

OTOH the SimpleKey will delegate equals/hashCode to the parameters, so the key attribute is needed if one of the parameters doesn't implement them (or cannot be serialized, or is just too big to be stored in the cache).

File: documentation/src/main/asciidoc/stories/assembly_hibernate_cache.adoc (hunk 1 above)
Comment (oraNod) [reply documentation/src/main/asciidoc/stories/assembly_hibernate_cache.adoc:8]: I lupdated the link to "current"

File: documentation/src/main/asciidoc/topics/con_spring_cache_annotations.adoc (hunk 3 above)
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/con_spring_cache_annotations.adoc]: @danberindei I updated this note to drop the detail about Spring behaviour and included that link in your comment. Could you please take another look? Thanks for the review!

**Comment by karesti** [comment]
@oraNod link me to the backport when the merge is possible

**Pull request merged**